R/zchunk_LA114.wind-energy.R

```diff
@@ -28,43 +30,237 @@ module_energy_LA114.wind_DISABLED <- function(command, ...) {
+        rename(mid.price = `mid-price`, curve.exponent = `curve-exponent`, base.price = `base-price`) %>% # Modifying title names
+        # Romania is listed in the Zhou file as rou and in the iso_GCAM_regID file as rom. rou will be changed to rom.
+        mutate(iso = replace(iso, iso == "rou", "rom")) %>%
+        left_join_error_no_match(iso_GCAM_regID, by = "iso") %>% # Attach region IDs to the countries
+        # Convert prices per unit of energy from 2007 USD/kWh to 1975 USD/GJ
+        mutate(mid.price = mid.price * gdp_deflator(1975, 2007) / CONV_KWH_GJ,
+               base.price = base.price * gdp_deflator(1975, 2007) / CONV_KWH_GJ) ->
+        L114.RsrcCurves_EJ_ctry_wind
+
+    } # End of old system behavior function
+
+    # Evaluate all supply curves by GCAM_region_ID to generate (price, supply) points
+    # All countries within each GCAM region ID will share the same price axis. The minimum and maximum prices
+         # will be from the countries within the region with the lowest minimum and highest maximum prices, respectively.
+         # The minimum prices are taken from the Zhou data. The maximum prices are calculated using the function,
+         # generate_max_prices, which will determine the price at which supply is 95% of the max resource.
+
+    # The price increment between min and max price is set at 0.1 USD
+    L114.price_inc <- 0.1 # 1975 USD
+
+
+    if(OLD_DATA_SYSTEM_BEHAVIOR) {
+
+      # The output of this old code is slightly different than the new method. The difference is the calculation of the maximum price.
+           # In the new method, the maximum price is determined to be the price at which supply is 95% of the max resource.
+           # In the old method, the maximum price is the price at which supply does not change more than 1%, evaluated from
+           # low to high price (at least 10 dollars more than minimum price, i.e., 100 min calculations at 0.1 increments).
+      L114.supply_tol <- 0.01
+      regional_price_supply_points <- tibble()
+      for(region_ID in unique(L114.RsrcCurves_EJ_ctry_wind$GCAM_region_ID)) {
+        L114.curr_region_RsrcCurves <- subset(L114.RsrcCurves_EJ_ctry_wind, GCAM_region_ID == region_ID)
+        L114.price <- min(L114.curr_region_RsrcCurves$base.price)
+        # Evaluate the supply curve until the supply doesn't seem to be changing more than a tolerance
+        # with a minimum number of evaluations to ensure we make progress into a very shallow curve as well.
+        L114.prev_supply <- 0
+        L114.min_calcs <- 100
+        repeat {
+          # Evaluate supply curves for all coutries in the region at the current price and sum up the supply
+          L114.SupplyPoints.currR <- tibble(GCAM_region_ID=region_ID, price=L114.price, supply=0)
+          for(i in seq_len(nrow(L114.curr_region_RsrcCurves))) {
+            L114.SupplyPoints.currR$supply <- L114.SupplyPoints.currR$supply + do.call(evaluate_smooth_res_curve,
```

Comment:
This inner loop is executing over 39,000 times, and `do.call` is quite expensive. I couldn't see any reason you guys were using it, so changed it (in ca0dca6) to a straight call to `evaluate_smooth_res_curve`. This drops the chunk's execution time from 23 to 7 seconds! It remains one of the slowest in the data system, but at least that's not completely crazy :)